### PR TITLE
Move all communication operations from collective

### PR DESCRIPTION
**Pull request by LiYuRio**
### PR types

Others

### PR changes

Others

### Describe

Move all communicated operations from collective.py to communication. Make the communication API under distributed reuse the new implementation under stream module.

- [x] ReduceOp
- [X] all_reduce
- [X] alltoall
- [X] alltoall_single
- [X] broadcast
- [X] reduce
- [X] send
- [X] scatter
- [X]  isend
- [X] recv
- [X] irecv
- [X] batch_isend_irecv
- [X] P2POp
- [X] reduce_scatter
- [X] is_initialized
- [X] destroy_process_group
- [X] get_group
- [x] all_gather
- [ ] all_gather_object
- [ ] wait
- [ ] barrier
- [ ] new_group

Other remain APIs will be moved in next PRs.

### CODE REVIEW

File: python/paddle/distributed/communication/stream/recv.py [hunk 1]
```diff
@@ -13,21 +13,43 @@
...
 
     return task
 
 
+def _recv_in_static_mode(tensor, src_rank_in_group, group, sync_op,
+                         use_calc_stream):
+    op_type = 'recv_v2'
+    data_feeder.check_variable_and_dtype(
+        tensor, 'tensor', ['float16', 'float32', 'float64', 'int32', 'int64'],
+        'recv')
```
Comment (HermitSun) [inline python/paddle/distributed/communication/stream/recv.py]: 1. 数据类型是不是还没加上（int8、uint8、bool，我看 all_reduce 写了）
2. 静态图支持 bf16 吗🤔

File: python/paddle/distributed/communication/stream/send.py [hunk 2]
```diff
@@ -13,21 +13,42 @@
...
 
     return task
 
 
+def _send_in_static_mode(tensor, dst_rank_in_group, group, sync_op,
+                         use_calc_stream):
+    op_type = 'send_v2'
+    data_feeder.check_variable_and_dtype(
+        tensor, 'tensor', ['float16', 'float32', 'float64', 'int32', 'int64'],
+        'send')
```
Comment (HermitSun) [inline python/paddle/distributed/communication/stream/send.py]: 同上

File: python/paddle/distributed/communication/stream/recv.py (hunk 1 above)
Comment (LiYuRio) [reply python/paddle/distributed/communication/stream/recv.py]: 数据类型后面单独搞一个pr专门加吧，这个pr主要是把collective里的东西移到communication下，然后复用stream下的API

File: python/paddle/distributed/communication/group.py [hunk 3]
```diff
@@ -94,3 +96,25 @@ def _add_new_group(group):
             "The group with id {} already exist.".format(group.id)
         )
     _GroupManager.group_map_by_id[group.id] = group
+
+
+def _warn_cur_rank_not_in_group(group):
+    global_rank = dist.get_rank()
```
Comment (HermitSun) [inline python/paddle/distributed/communication/group.py:111]: 是不是少了个`import paddle.distributed as dist`🤔

File: python/paddle/distributed/communication/stream/broadcast.py [hunk 4]
```diff
@@ -13,21 +13,63 @@
...
-    task = group.process_group.broadcast(tensor, src, sync_op)
+    task = group.process_group.broadcast(tensor, src_rank_in_group, sync_op)
     if sync_op:
         task.wait()
 
     return task
 
 
+def _broadcast_in_static_mode(
+    tensor, src_rank_in_group, group, sync_op, use_calc_stream
```
Comment (HermitSun) [inline python/paddle/distributed/communication/stream/broadcast.py:41]: 这里的`group`参数是不是没用到🤔我看底下调用的时候只传了4个参数

Comment (LiYuRio) [reply python/paddle/distributed/communication/stream/broadcast.py:41]: 写漏了

File: python/paddle/distributed/communication/group.py (hunk 3 above)
Comment (LiYuRio) [reply python/paddle/distributed/communication/group.py:111]: 还是写漏了

**Review by sljlp — APPROVED** [review]
LGTM

**Review by HermitSun — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**